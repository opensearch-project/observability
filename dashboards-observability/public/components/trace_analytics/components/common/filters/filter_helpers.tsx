/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiComboBox,
  EuiFieldText,
  EuiFormControlLayoutDelimited,
  EuiFormRow,
  EuiSpacer,
} from '@elastic/eui';
import _ from 'lodash';
import { TraceAnalyticsMode } from 'public/components/trace_analytics/home';
import React from 'react';

const getFields = (mode: TraceAnalyticsMode, page: 'dashboard' | 'traces' | 'services' | 'app') =>
  (mode === 'data_prepper' ? {
    dashboard: ['traceGroup', 'serviceName', 'error', 'status.message', 'latency'],
    traces: ['traceId', 'traceGroup', 'serviceName', 'error', 'status.message', 'latency'],
    services: ['traceGroup', 'serviceName', 'error', 'status.message', 'latency'],
    app: ['traceId', 'traceGroup', 'serviceName'],
  }[page] : {
    dashboard: ['process.serviceName', 'error', 'latency'],
    traces: ['traceID', 'operationName', 'process.serviceName', 'error', 'latency'],
    services: ['process.serviceName', 'error', 'latency'],
    app: ['traceID', 'process.serviceName'],
  }[page]);
// filters will take effect and can be manually added
export const getFilterFields = (mode: TraceAnalyticsMode, page: 'dashboard' | 'traces' | 'services' | 'app') => getFields(mode, page);
// filters will take effect
export const getValidFilterFields = (mode: TraceAnalyticsMode, page: 'dashboard' | 'traces' | 'services' | 'app') => {
  const fields = getFields(mode, page);
  if (page !== 'services') return [...fields, 'Latency percentile within trace group'];
  return fields;
};

const getType = (field: string): string | null => {
  const typeMapping = {
    attributes: {
      host: {
        port: 'long',
      },
      http: {
        response_content_length: 'long',
        status_code: 'long',
      },
      net: {
        port: 'long',
      },
    },
    attributes_host: {
      port: 'long',
    },
    attributes_http: {
      response_content_length: 'long',
      status_code: 'long',
    },
    attributes_net: {
      port: 'long',
    },
    durationInNanos: 'long',
    latency: 'long',
    endTime: 'date_nanos',
    startTime: 'date_nanos',
  };
  const type = _.get(typeMapping, field, 'keyword');
  return typeof type === 'string' ? type : null;
};

export const getInvertedOperator = (operator: string, inverted: boolean) => {
  if (operator.includes('between')) return inverted ? 'is not between' : 'is between';
  else if (operator.includes('exist')) return inverted ? 'does not exist' : 'exists';
  else if (operator === 'is' || operator === 'is not') return inverted ? 'is not' : 'is';
};

export const getOperatorOptions = (field: string) => {
  const type = getType(field);
  const operatorMapping = {
    long: [
      {
        label: 'is between',
      },
      {
        label: 'is not between',
      },
    ],
    date_nanos: [
      {
        label: 'is between',
      },
      {
        label: 'is not between',
      },
    ],
    keyword: [],
    default_first: [
      {
        label: 'is',
      },
      {
        label: 'is not',
      },
    ],
    default_last: [
      {
        label: 'exists',
      },
      {
        label: 'does not exist',
      },
    ],
  };
  const operators = [
    ...operatorMapping.default_first,
    ..._.get(operatorMapping, type),
    ...operatorMapping.default_last,
  ];
  return operators;
};

export const getValueComponent = (
  field: string,
  operator: string,
  value: any,
  setValue: (v: any) => void
) => {
  const textField = (
    <>
      <EuiSpacer size="s" />
      <EuiFormRow label={'Value'}>
        <EuiFieldText
          placeholder="Enter a value"
          value={value}
          onChange={(e) => setValue(e.target.value)}
        />
      </EuiFormRow>
    </>
  );

  const getRangeField = () => {
    const getFromValue = () => {
      if (value?.from) {
        return value.from.includes('\u221E') ? '' : value.from;
      }
      return '';
    };
    const getToValue = () => {
      if (value?.to) {
        return value.to.includes('\u221E') ? '' : value.to;
      }
      return '';
    };
    const setFromValue = (from: string) => {
      setValue({ from: from || '-\u221E', to: getToValue() || '\u221E' });
    };
    const setToValue = (to: string) => {
      setValue({ from: getFromValue() || '-\u221E', to: to || '\u221E' });
    };
    return (
      <>
        <EuiSpacer size="s" />
        <EuiFormRow label={'Value'}>
          <EuiFormControlLayoutDelimited
            startControl={
              <input
                type="string"
                placeholder="Start of range"
                className="euiFieldText"
                value={getFromValue()}
                onChange={(e) => setFromValue(e.target.value)}
              />
            }
            endControl={
              <input
                type="string"
                placeholder="End of range"
                className="euiFieldText"
                value={getToValue()}
                onChange={(e) => setToValue(e.target.value)}
              />
            }
          />
        </EuiFormRow>
      </>
    );
  };

  const getComboBoxField = () => {
    return (
      <>
        <EuiSpacer size="s" />
        <EuiFormRow label={'Value'}>
          <EuiComboBox
            placeholder="Select a value"
            options={[
              {
                label: 'true',
              },
              {
                label: 'false',
              },
            ]}
            onChange={setValue}
            selectedOptions={value || []}
            singleSelection={true}
          />
        </EuiFormRow>
      </>
    );
  };

  if (field === 'error' && (operator === 'is' || operator === 'is not')) {
    return getComboBoxField();
  }

  const valueMapping = {
    is: textField,
    'is not': textField,
    'is between': getRangeField(),
    'is not between': getRangeField(),
    exists: null,
    'does not exist': null,
  };

  return valueMapping[operator];
};
