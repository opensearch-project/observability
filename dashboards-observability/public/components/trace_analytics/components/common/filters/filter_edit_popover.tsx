/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiButton,
  EuiButtonEmpty,
  EuiComboBox,
  EuiComboBoxOptionOption,
  EuiFlexGroup,
  EuiFlexItem,
  EuiFormRow,
  EuiSpacer,
} from '@elastic/eui';
import React, { useState } from 'react';
import { FilterType } from './filters';
import { getInvertedOperator, getOperatorOptions, getValueComponent } from './filter_helpers';

export function FilterEditPopover(props: {
  filter?: FilterType;
  index: number;
  setFilter: (newFilter: FilterType, index: number) => void;
  closePopover: () => void;
  filterFieldOptions: Array<{ label: string }>;
}) {
  const [selectedFieldOptions, setSelectedFieldOptions] = useState<
    Array<EuiComboBoxOptionOption<string>>
  >(props.filter ? [{ label: props.filter.field }] : []);
  const [selectedOperatorOptions, setSelectedOperatorOptions] = useState<
    Array<EuiComboBoxOptionOption<string>>
  >(
    props.filter
      ? [{ label: getInvertedOperator(props.filter.operator, props.filter.inverted) }]
      : []
  );
  const [filterValue, setFilterValue] = useState(props.filter?.value || '');

  return (
    <div style={{ width: 400 }}>
      {/* invisible button workaround to prevent auto focus on context menu panel switch */}
      <button style={{ width: 0, height: 0, position: 'fixed', marginLeft: -1000, bottom: 0 }} />
      <EuiFlexGroup gutterSize="s">
        <EuiFlexItem grow={6}>
          <EuiFormRow label={'Field'}>
            <EuiComboBox
              placeholder="Select a field first"
              isClearable={false}
              options={props.filterFieldOptions}
              selectedOptions={selectedFieldOptions}
              onChange={(e) => {
                setSelectedFieldOptions(e);
                setSelectedOperatorOptions([]);
                setFilterValue('');
              }}
              singleSelection={{ asPlainText: true }}
            />
          </EuiFormRow>
        </EuiFlexItem>
        <EuiFlexItem grow={5}>
          <EuiFormRow label={'Operator'}>
            <EuiComboBox
              placeholder={selectedFieldOptions.length === 0 ? 'Waiting' : 'Select'}
              isClearable={false}
              isDisabled={selectedFieldOptions.length === 0}
              options={
                selectedFieldOptions.length === 0
                  ? []
                  : getOperatorOptions(selectedFieldOptions[0].label)
              }
              selectedOptions={selectedOperatorOptions}
              onChange={(e) => {
                setSelectedOperatorOptions(e);
                setFilterValue('');
              }}
              singleSelection={{ asPlainText: true }}
            />
          </EuiFormRow>
        </EuiFlexItem>
      </EuiFlexGroup>
      {selectedOperatorOptions.length > 0 &&
        getValueComponent(
          selectedFieldOptions[0].label,
          selectedOperatorOptions[0].label,
          filterValue,
          setFilterValue
        )}
      <EuiSpacer size="m" />
      <EuiFlexGroup gutterSize="s" justifyContent="flexEnd">
        <EuiFlexItem grow={false}>
          <EuiButtonEmpty
            data-test-subj="filter-popover-cancel-button"
            onClick={props.closePopover}
          >
            Cancel
          </EuiButtonEmpty>
        </EuiFlexItem>
        <EuiFlexItem grow={false}>
          <EuiButton
            fill
            disabled={
              selectedFieldOptions.length === 0 ||
              selectedOperatorOptions.length === 0 ||
              (filterValue.length === 0 && !selectedOperatorOptions[0]?.label?.includes('exist'))
            }
            onClick={() => {
              props.closePopover();
              props.setFilter(
                {
                  field: selectedFieldOptions[0].label,
                  operator: selectedOperatorOptions[0].label,
                  value: selectedOperatorOptions[0].label.includes('exist')
                    ? 'exists'
                    : filterValue,
                  inverted: selectedOperatorOptions[0].label.includes('not'),
                  disabled: props.filter ? props.filter.disabled : false,
                },
                props.index
              );
            }}
          >
            Save
          </EuiButton>
        </EuiFlexItem>
      </EuiFlexGroup>
    </div>
  );
}
