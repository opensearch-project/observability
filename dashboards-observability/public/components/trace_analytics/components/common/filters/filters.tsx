/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiBadge,
  EuiButtonEmpty,
  EuiButtonIcon,
  EuiContextMenu,
  EuiContextMenuPanelDescriptor,
  EuiFlexGroup,
  EuiFlexItem,
  EuiIcon,
  EuiPopover,
  EuiPopoverTitle,
  EuiTextColor,
} from '@elastic/eui';
import React, { useMemo, useState } from 'react';
import { FilterEditPopover } from './filter_edit_popover';
import { getFilterFields, getValidFilterFields } from './filter_helpers';

export interface FilterType {
  field: string;
  operator: string;
  value: any;
  inverted: boolean;
  disabled: boolean;
  custom?: any;
  locked?: boolean;
}

export interface FiltersProps {
  filters: FilterType[];
  appConfigs?: FilterType[];
  setFilters: (filters: FilterType[]) => void;
}

interface FiltersOwnProps extends FiltersProps {
  page: 'dashboard' | 'traces' | 'services' | 'app';
}

export function Filters(props: FiltersOwnProps) {
  // set a filter at an index. if newFilter doesn't exist, remove filter at the index
  // if index doesn't exist, append newFilter to the end
  const setFilter = (newFilter: FilterType, index: number) => {
    const newFilters = [...props.filters];
    if (newFilter) newFilters.splice(index, 1, newFilter);
    else newFilters.splice(index, 1);
    props.setFilters(newFilters);
  };

  const validFilterFields = useMemo(() => getValidFilterFields(props.page), [props.page]);
  const filterFieldOptions = useMemo(
    () => getFilterFields(props.page).map((field) => ({ label: field })),
    [props.page]
  );

  const globalPopoverPanels = [
    {
      id: 0,
      title: 'Change all filters',
      items: [
        {
          name: 'Enable all',
          icon: <EuiIcon type="eye" size="m" />,
          onClick: () => {
            props.setFilters(
              props.filters.map((filter) => ({
                ...filter,
                disabled: filter.locked ? filter.disabled : false,
              }))
            );
          },
        },
        {
          name: 'Disable all',
          icon: <EuiIcon type="eyeClosed" size="m" />,
          onClick: () => {
            props.setFilters(
              props.filters.map((filter) => ({
                ...filter,
                disabled: filter.locked ? filter.disabled : true,
              }))
            );
          },
        },
        {
          name: 'Invert inclusion',
          icon: <EuiIcon type="invert" size="m" />,
          onClick: () => {
            props.setFilters(
              // if filter.custom.query exists, it's a customized filter and "inverted" is alwasy false
              props.filters.map((filter) => ({
                ...filter,
                inverted: filter.locked
                  ? filter.inverted
                  : filter.custom?.query
                  ? false
                  : !filter.inverted,
              }))
            );
          },
        },
        {
          name: 'Invert enabled/disabled',
          icon: <EuiIcon type="eye" size="m" />,
          onClick: () => {
            props.setFilters(
              props.filters.map((filter) => ({
                ...filter,
                disabled: filter.locked ? filter.disabled : !filter.disabled,
              }))
            );
          },
        },
        {
          name: 'Remove all',
          icon: <EuiIcon type="trash" size="m" />,
          onClick: () => {
            props.setFilters([]);
          },
        },
      ],
    },
  ];

  const getFilterPopoverPanels = (
    filter: FilterType,
    index: number,
    closePopover: () => void
  ): EuiContextMenuPanelDescriptor[] => [
    {
      id: 0,
      items: [
        {
          name: 'Edit filter',
          icon: <EuiIcon type="invert" size="m" />,
          disabled: !!filter.custom?.query || validFilterFields.indexOf(filter.field) === -1,
          panel: 1,
        },
        {
          name: `${filter.inverted ? 'Include' : 'Exclude'} results`,
          icon: <EuiIcon type={filter.inverted ? 'plusInCircle' : 'minusInCircle'} size="m" />,
          disabled: !!filter.custom?.query || validFilterFields.indexOf(filter.field) === -1,
          onClick: () => {
            filter.inverted = !filter.inverted;
            setFilter(filter, index);
          },
        },
        {
          name: filter.disabled ? 'Re-enable' : 'Temporarily disable',
          icon: <EuiIcon type={filter.disabled ? 'eye' : 'eyeClosed'} size="m" />,
          disabled: validFilterFields.indexOf(filter.field) === -1,
          onClick: () => {
            filter.disabled = !filter.disabled;
            setFilter(filter, index);
          },
        },
        {
          name: 'Delete',
          icon: <EuiIcon data-test-subj="deleteFilterIcon" type="trash" size="m" />,
          onClick: () => setFilter(null, index),
        },
      ],
    },
    {
      id: 1,
      width: 430,
      title: 'Edit filter',
      content: (
        <div style={{ margin: 15 }}>
          <FilterEditPopover
            filterFieldOptions={filterFieldOptions}
            filter={filter}
            index={index}
            setFilter={setFilter}
            closePopover={closePopover}
          />
        </div>
      ),
    },
  ];

  const GlobalFilterButton = () => {
    const [isPopoverOpen, setIsPopoverOpen] = useState(false);
    return (
      <EuiPopover
        isOpen={isPopoverOpen}
        closePopover={() => setIsPopoverOpen(false)}
        button={
          <EuiButtonIcon
            onClick={() => setIsPopoverOpen(true)}
            iconType="filter"
            title="Change all filters"
            aria-label="Change all filters"
          />
        }
        anchorPosition="rightUp"
        panelPaddingSize="none"
        withTitle
      >
        <EuiContextMenu initialPanelId={0} panels={globalPopoverPanels} />
      </EuiPopover>
    );
  };

  const AddFilterButton = () => {
    const [isPopoverOpen, setIsPopoverOpen] = useState(false);
    const button = (
      <EuiButtonEmpty
        size="xs"
        onClick={() => {
          setIsPopoverOpen(true);
        }}
      >
        + Add filter
      </EuiButtonEmpty>
    );

    return (
      <EuiPopover
        button={button}
        isOpen={isPopoverOpen}
        closePopover={() => setIsPopoverOpen(false)}
        anchorPosition="downLeft"
        withTitle
      >
        <EuiPopoverTitle>{'Add filter'}</EuiPopoverTitle>
        <FilterEditPopover
          filterFieldOptions={filterFieldOptions}
          index={props.filters.length}
          setFilter={setFilter}
          closePopover={() => setIsPopoverOpen(false)}
        />
      </EuiPopover>
    );
  };

  const renderFilters = () => {
    const FilterBadge = ({ filter, index }: { filter: FilterType; index: number }) => {
      const [isPopoverOpen, setIsPopoverOpen] = useState(false);
      const disabled = filter.locked || filter.disabled;
      const className =
        'globalFilterItem' +
        (disabled ? ' globalFilterItem-isDisabled' : '') +
        (filter.inverted ? ' globalFilterItem-isExcluded' : '');
      const value =
        typeof filter.value === 'string'
          ? filter.value
          : Array.isArray(filter.value)  // combo box
          ? filter.value[0].label
          : `${filter.value.from} to ${filter.value.to}`;  // range selector
      const filterLabel = filter.inverted ? (
        <>
          <EuiTextColor color={disabled ? 'default' : 'danger'}>{'NOT '}</EuiTextColor>
          <EuiTextColor color="default">{`${filter.field}: ${value}`}</EuiTextColor>
        </>
      ) : (
        `${filter.field}: ${value}`
      );

      const badge = (
        <EuiBadge
          className={className}
          onClick={() => setIsPopoverOpen(true)}
          onClickAriaLabel="Open filter settings"
          color={disabled ? '#e7e9f0' : 'hollow'}
          iconType="cross"
          iconSide="right"
          iconOnClick={() => {
            setFilter(null, index);
          }}
          iconOnClickAriaLabel="Remove filter"
          data-test-subj="filterBadge"
        >
          {filterLabel}
        </EuiBadge>
      );
      return (
        <EuiFlexItem grow={false} key={`filter-${index}`}>
          <EuiPopover
            data-test-subj={`${filter.value}FilterBadge`}
            isOpen={isPopoverOpen}
            closePopover={() => setIsPopoverOpen(false)}
            panelPaddingSize="none"
            button={badge}
          >
            <EuiContextMenu
              initialPanelId={0}
              panels={getFilterPopoverPanels(filter, index, () => setIsPopoverOpen(false))}
            />
          </EuiPopover>
        </EuiFlexItem>
      );
    };

    return (
      <>
        {props.filters.length > 0
          ? props.filters.map((filter, i) => <FilterBadge filter={filter} index={i} key={i} />)
          : null}
      </>
    );
  };

  const filterComponents = useMemo(() => renderFilters(), [props.filters]);

  return (
    <EuiFlexGroup gutterSize="xs" alignItems="center" responsive={false} style={{ minHeight: 32 }}>
      <EuiFlexItem grow={false}>
        <GlobalFilterButton />
      </EuiFlexItem>
      {filterComponents}
      <EuiFlexItem grow={false}>
        <AddFilterButton />
      </EuiFlexItem>
    </EuiFlexGroup>
  );
}
