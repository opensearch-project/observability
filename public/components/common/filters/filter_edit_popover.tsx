/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import {
  EuiButton,
  EuiButtonEmpty,
  EuiComboBox,
  EuiComboBoxOptionOption,
  EuiFlexGroup,
  EuiFlexItem,
  EuiFormRow,
  EuiSpacer,
} from '@elastic/eui';
import React, { useState } from 'react';
import { FilterType } from './filters';
import { getInvertedOperator, getOperatorOptions, getValueComponent } from './filter_helpers';

export function FilterEditPopover(props: {
  filter?: FilterType;
  index: number;
  setFilter: (newFilter: FilterType, index: number) => void;
  closePopover: () => void;
  filterFieldOptions: Array<{ label: string }>;
}) {
  const [selectedFieldOptions, setSelectedFieldOptions] = useState<
    Array<EuiComboBoxOptionOption<string>>
  >(props.filter ? [{ label: props.filter.field }] : []);
  const [selectedOperatorOptions, setSelectedOperatorOptions] = useState<
    Array<EuiComboBoxOptionOption<string>>
  >(
    props.filter
      ? [{ label: getInvertedOperator(props.filter.operator, props.filter.inverted) }]
      : []
  );
  const [filterValue, setFilterValue] = useState(props.filter?.value || '');

  return (
    <div style={{ width: 400 }}>
      {/* invisible button workaround to prevent auto focus on context menu panel switch */}
      <button style={{ width: 0, height: 0, position: 'fixed', marginLeft: -1000, bottom: 0 }} />
      <EuiFlexGroup gutterSize="s">
        <EuiFlexItem grow={6}>
          <EuiFormRow label={'Field'}>
            <EuiComboBox
              placeholder="Select a field first"
              isClearable={false}
              options={props.filterFieldOptions}
              selectedOptions={selectedFieldOptions}
              onChange={(e) => {
                setSelectedFieldOptions(e);
                setSelectedOperatorOptions([]);
                setFilterValue('');
              }}
              singleSelection={{ asPlainText: true }}
            />
          </EuiFormRow>
        </EuiFlexItem>
        <EuiFlexItem grow={5}>
          <EuiFormRow label={'Operator'}>
            <EuiComboBox
              placeholder={selectedFieldOptions.length === 0 ? 'Waiting' : 'Select'}
              isClearable={false}
              isDisabled={selectedFieldOptions.length === 0}
              options={
                selectedFieldOptions.length === 0
                  ? []
                  : getOperatorOptions(selectedFieldOptions[0].label)
              }
              selectedOptions={selectedOperatorOptions}
              onChange={(e) => {
                setSelectedOperatorOptions(e);
                setFilterValue('');
              }}
              singleSelection={{ asPlainText: true }}
            />
          </EuiFormRow>
        </EuiFlexItem>
      </EuiFlexGroup>
      {selectedOperatorOptions.length > 0 &&
        getValueComponent(
          selectedFieldOptions[0].label,
          selectedOperatorOptions[0].label,
          filterValue,
          setFilterValue
        )}
      <EuiSpacer size="m" />
      <EuiFlexGroup gutterSize="s" justifyContent="flexEnd">
        <EuiFlexItem grow={false}>
          <EuiButtonEmpty
            data-test-subj="filter-popover-cancel-button"
            onClick={props.closePopover}
          >
            Cancel
          </EuiButtonEmpty>
        </EuiFlexItem>
        <EuiFlexItem grow={false}>
          <EuiButton
            fill
            disabled={
              selectedFieldOptions.length === 0 ||
              selectedOperatorOptions.length === 0 ||
              (filterValue.length === 0 && !selectedOperatorOptions[0]?.label?.includes('exist'))
            }
            onClick={() => {
              props.closePopover();
              props.setFilter(
                {
                  field: selectedFieldOptions[0].label,
                  operator: selectedOperatorOptions[0].label,
                  value: selectedOperatorOptions[0].label.includes('exist')
                    ? 'exists'
                    : filterValue,
                  inverted: selectedOperatorOptions[0].label.includes('not'),
                  disabled: props.filter ? props.filter.disabled : false,
                },
                props.index
              );
            }}
          >
            Save
          </EuiButton>
        </EuiFlexItem>
      </EuiFlexGroup>
    </div>
  );
}
