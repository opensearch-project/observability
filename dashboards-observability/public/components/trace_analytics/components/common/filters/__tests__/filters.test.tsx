/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { render } from '@testing-library/react';
import { configure, mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Filters } from '../filters';

describe('Filter component', () => {
  configure({ adapter: new Adapter() });

  it('renders filters', () => {
    const setFilters = jest.fn();
    const wrapper = mount(
      <Filters page="dashboard" filters={[]} setFilters={setFilters} appConfigs={[]} />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
