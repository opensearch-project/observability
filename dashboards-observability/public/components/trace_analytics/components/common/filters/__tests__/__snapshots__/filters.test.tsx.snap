// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter component renders filters 1`] = `
<Filters
  filters={Array []}
  page="dashboard"
  setFilters={[MockFunction]}
>
  <EuiFlexGroup
    alignItems="center"
    gutterSize="xs"
    responsive={false}
    style={
      Object {
        "minHeight": 32,
      }
    }
  >
    <div
      className="euiFlexGroup euiFlexGroup--gutterExtraSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow"
      style={
        Object {
          "minHeight": 32,
        }
      }
    >
      <EuiFlexItem
        grow={false}
      >
        <div
          className="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <GlobalFilterButton>
            <EuiPopover
              anchorPosition="rightUp"
              button={
                <EuiButtonIcon
                  aria-label="Change all filters"
                  iconType="filter"
                  onClick={[Function]}
                  title="Change all filters"
                />
              }
              closePopover={[Function]}
              display="inlineBlock"
              hasArrow={true}
              isOpen={false}
              ownFocus={false}
              panelPaddingSize="none"
              withTitle={true}
            >
              <EuiOutsideClickDetector
                isDisabled={true}
                onOutsideClick={[Function]}
              >
                <div
                  className="euiPopover euiPopover--anchorRightUp euiPopover--withTitle"
                  onKeyDown={[Function]}
                  onMouseDown={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchStart={[Function]}
                >
                  <div
                    className="euiPopover__anchor"
                  >
                    <EuiButtonIcon
                      aria-label="Change all filters"
                      iconType="filter"
                      onClick={[Function]}
                      title="Change all filters"
                    >
                      <button
                        aria-label="Change all filters"
                        className="euiButtonIcon euiButtonIcon--primary"
                        disabled={false}
                        onClick={[Function]}
                        title="Change all filters"
                        type="button"
                      >
                        <EuiIcon
                          aria-hidden="true"
                          className="euiButtonIcon__icon"
                          size="m"
                          type="filter"
                        >
                          <EuiIconEmpty
                            aria-hidden={true}
                            className="euiIcon euiIcon--medium euiIcon-isLoading euiButtonIcon__icon"
                            focusable="false"
                            role="img"
                            style={null}
                          >
                            <svg
                              aria-hidden={true}
                              className="euiIcon euiIcon--medium euiIcon-isLoading euiButtonIcon__icon"
                              focusable="false"
                              height={16}
                              role="img"
                              style={null}
                              viewBox="0 0 16 16"
                              width={16}
                              xmlns="http://www.w3.org/2000/svg"
                            />
                          </EuiIconEmpty>
                        </EuiIcon>
                      </button>
                    </EuiButtonIcon>
                  </div>
                </div>
              </EuiOutsideClickDetector>
            </EuiPopover>
          </GlobalFilterButton>
        </div>
      </EuiFlexItem>
      <EuiFlexItem
        grow={false}
      >
        <div
          className="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <AddFilterButton>
            <EuiPopover
              anchorPosition="downLeft"
              button={
                <EuiButtonEmpty
                  onClick={[Function]}
                  size="xs"
                >
                  + Add filter
                </EuiButtonEmpty>
              }
              closePopover={[Function]}
              display="inlineBlock"
              hasArrow={true}
              isOpen={false}
              ownFocus={false}
              panelPaddingSize="m"
              withTitle={true}
            >
              <EuiOutsideClickDetector
                isDisabled={true}
                onOutsideClick={[Function]}
              >
                <div
                  className="euiPopover euiPopover--anchorDownLeft euiPopover--withTitle"
                  onKeyDown={[Function]}
                  onMouseDown={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchStart={[Function]}
                >
                  <div
                    className="euiPopover__anchor"
                  >
                    <EuiButtonEmpty
                      onClick={[Function]}
                      size="xs"
                    >
                      <button
                        className="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--xSmall"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        <EuiButtonContent
                          className="euiButtonEmpty__content"
                          iconSide="left"
                          textProps={
                            Object {
                              "className": "euiButtonEmpty__text",
                            }
                          }
                        >
                          <span
                            className="euiButtonContent euiButtonEmpty__content"
                          >
                            <span
                              className="euiButtonEmpty__text"
                            >
                              + Add filter
                            </span>
                          </span>
                        </EuiButtonContent>
                      </button>
                    </EuiButtonEmpty>
                  </div>
                </div>
              </EuiOutsideClickDetector>
            </EuiPopover>
          </AddFilterButton>
        </div>
      </EuiFlexItem>
    </div>
  </EuiFlexGroup>
</Filters>
`;
